cli/cli/helpers/grafloki/kubernetes_grafloki.go

```diff
@@ -0,0 +1,477 @@
+			if err := k8sManager.RemoveNamespace(ctx, glNamespace); err != nil {
+				logrus.Warnf("Attempted to remove namespace '%v' after an error occurred creating it but an error occurred removing it.", graflokiNamespace)
+				logrus.Warnf("Manually remove namepsace with name: %v", glNamespace.Name)
+			}
+		}
+	}()
+	if err != nil {
+		return "", stacktrace.Propagate(err, "An error occurred creating namespace '%v'", graflokiNamespace)
+	}
+
+	lokiDeployment, err := k8sManager.CreateDeployment(
+		ctx,
+		graflokiNamespace,
+		lokiDeploymentName,
+		lokiLabels,
+		map[string]string{},
+		[]apiv1.Container{}, // no init containers
+		[]apiv1.Container{
+			{
+				Name:  "loki",
+				Image: lokiImage,
+				Ports: []apiv1.ContainerPort{
+					{
+						Name:          "",
+						HostPort:      0,
+						ContainerPort: lokiPort,
+						Protocol:      "",
+						HostIP:        "",
+					},
+				},
+				Command:    nil,
+				Args:       nil,
+				WorkingDir: "",
+				EnvFrom:    nil,
+				Env:        nil,
+				Resources: apiv1.ResourceRequirements{
+					Limits:   nil,
+					Requests: nil,
+					Claims:   nil,
+				},
+				ResizePolicy:  nil,
+				VolumeMounts:  nil,
+				VolumeDevices: nil,
+				LivenessProbe: nil,
+				ReadinessProbe: &apiv1.Probe{
+					ProbeHandler: apiv1.ProbeHandler{
+						Exec: nil,
+						HTTPGet: &apiv1.HTTPGetAction{
+							Path:        "/ready",
+							Port:        intstr.FromInt(lokiPort),
+							Host:        "",
+							Scheme:      "",
+							HTTPHeaders: nil,
+						},
+						TCPSocket: nil,
+						GRPC:      nil,
+					},
+					InitialDelaySeconds:           lokiProbInitialDelaySeconds,
+					TimeoutSeconds:                lokiProbeTimeoutSeconds,
+					PeriodSeconds:                 lokiProbePeriodSeconds,
+					SuccessThreshold:              0,
+					FailureThreshold:              0,
+					TerminationGracePeriodSeconds: nil,
+				},
+				StartupProbe:             nil,
+				Lifecycle:                nil,
+				TerminationMessagePath:   "",
+				TerminationMessagePolicy: "",
+				ImagePullPolicy:          "",
+				SecurityContext:          nil,
+				Stdin:                    false,
+				StdinOnce:                false,
+				TTY:                      false,
+			},
+		},
+		[]apiv1.Volume{},
+		&apiv1.Affinity{
+			NodeAffinity:    nil,
+			PodAffinity:     nil,
+			PodAntiAffinity: nil,
+		})
+	if err != nil {
+		return "", stacktrace.Propagate(err, "An error occurred creating Loki deployment.")
+	}
+	shouldRemoveLokiDeployment := false
+	defer func() {
+		if shouldRemoveLokiDeployment {
+			if err := k8sManager.RemoveDeployment(ctx, graflokiNamespace, lokiDeployment); err != nil {
+				logrus.Warnf("Attempted to remove Loki deployment after an error occurred creating it but an error occurred removing it.")
+				logrus.Warnf("Manually remove Loki deployment with name: %v", lokiDeployment.Name)
+			}
+		}
+	}()
+	logrus.Infof("Waiting for Loki deployment to come online (can take around 30s)... ")
```

Comment:
Sounds like Loki takes around 30s, and we're giving 40s. That seems like a pretty tight timeline, esp since I'm guessing that you're using your (super highpowered) Mac as benchmarking. What happens on a user's computer that's older or slower or more overloaded?

In Clear Thinking Shane Parrish talks about how a good rule-of-thumb for margin of safety is 2x what you think; might be useful waiting at least a minute for Loki to come online

Reply:
If there's a fear around too long of a wait in an infini-failed state, I thiiiiink the Kubernetes probes might have ways to detect this and eject early (maybe worth looking into)

Reply:
we encourage loki to ship faster, hence the tight timeline